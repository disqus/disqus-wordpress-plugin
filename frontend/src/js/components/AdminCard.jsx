import React from 'react';

const AdminCard = props =>
    <div className="card">
        {props.title ?
            <h2 className="title">
                {props.title}
            </h2>
        : null}
        {props.children}
    </div>;

export default AdminCard;
