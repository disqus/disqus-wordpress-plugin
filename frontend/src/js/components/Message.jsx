import React from 'react';

const Message = props =>
    <div className={`notice notice-${props.type} inline is-dismissible`}>
        <p>{props.children}</p>
    </div>;

export default Message;
