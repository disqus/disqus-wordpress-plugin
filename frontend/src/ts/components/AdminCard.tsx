import * as React from 'react';

export interface AdminCardProps {
    children?: React.ReactChildren,
    title: string,
}

const AdminCard = (props: AdminCardProps) =>
    <div className="card">
        {props.title ?
            <h2 className="title">
                {props.title}
            </h2>
        : null}
        {props.children}
    </div>;

export default AdminCard;
