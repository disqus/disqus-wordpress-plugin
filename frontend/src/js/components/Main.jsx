import React from 'react';
import { getWordpressAdminUrl } from 'utils';
import { restGet, restPut } from 'rest';
import Admin from 'components/Admin';
import Install from 'components/Install';
import Loading from 'components/Loading';
import Message from 'components/Message';

class Main extends React.Component {
    constructor() {
        super();

        this.state = {
            adminOptions: null,
            localShortname: '',
            localPublicKey: '',
            localSecretKey: '',
            localAdminAccessToken: '',
            localSSOButton: '',
            localSSOEnabled: false,
            localManualSync: false,
            message: null,
            isSiteFormLocked: true,
        };
    }

    componentDidMount() {
        restGet('settings', this.handleAdminOptionsResponse.bind(this));
    }

    saveOptions(data) {
        restPut('settings', data, response => {
            this.handleAdminOptionsResponse(response);
            this.addMessage(__('Saved changes successfully.'));
        });
    }

    handleAdminOptionsResponse(response) {
        if (!response)
            return;

        if (response.code !== 'OK') {
            this.addMessage(response.message, 'error');
            return;
        }

        const adminOptions = response.data || {};
        this.setState({
            adminOptions,
            localShortname: adminOptions.disqus_forum_url || '',
            localPublicKey: adminOptions.disqus_public_key || '',
            localSecretKey: adminOptions.disqus_secret_key || '',
            localAdminAccessToken: adminOptions.disqus_admin_access_token || '',
            localSSOButton: adminOptions.disqus_sso_button || '',
            localSSOEnabled: adminOptions.disqus_sso_enabled || false,
            localManualSync: adminOptions.disqus_manual_sync || false,
        });
    }

    handleInputChanged(key, evt) {
        const isCheckbox = evt.target.type === 'checkbox';
        this.setState({
            [key]: isCheckbox ? evt.target.checked : evt.target.value,
        });
    }

    handleSubmitForm(evt) {
        evt.preventDefault();

        var fields = {};
        Object.keys(this.state).forEach(currentIdKey => {
            if (currentIdKey in evt.target.elements)
                fields[evt.target.elements[currentIdKey].name] = this.state[currentIdKey];
        });

        this.saveOptions(fields);
    }

    addMessage(messageText, type) {
        this.setState({
            message: {
                text: messageText,
                type: type,
            },
        });
    }

    handleToggleState(stateKey) {
        this.setState({
            [stateKey]: !this.state[stateKey]
        });
    }

    renderMainView() {
        if (!window.DISQUS_WP.permissions.canManageSettings) {
            return __('You don\'t have permission to make any changes here. Please contact the site administrator to get access.');
        } else if (!this.state.adminOptions) {
            return <Loading />;
        } else if (!this.state.adminOptions.disqus_forum_url) {
            return (
                <Install
                    {...this.state}
                    onInputChange={this.handleInputChanged.bind(this)}
                    onSubmit={this.handleSubmitForm.bind(this)}
                    onToggleState={this.handleToggleState.bind(this)}
                />
            );
        }

        return (
            <Admin
                {...this.state}
                onInputChange={this.handleInputChanged.bind(this)}
                onSubmit={this.handleSubmitForm.bind(this)}
                onToggleState={this.handleToggleState.bind(this)}
            />
        );
    }

    render() {
        return (
            <div className="dsq-admin-wrapper">
                <div className="wrap">
                    <a href={getWordpressAdminUrl('disqus')} className="disqus-logo">
                        <img
                            src="https://a.disquscdn.com/dotcom/d-2407bda/img/brand/disqus-logo-blue-white.svg"
                            width={102}
                        />
                    </a>
                </div>
                <div className="wrap">
                    {this.state.message ?
                        <Message type={this.state.message.type}>{this.state.message.text}</Message>
                    : null}
                    {this.renderMainView()}
                </div>
            </div>
        );
    }
}

export default Main;
