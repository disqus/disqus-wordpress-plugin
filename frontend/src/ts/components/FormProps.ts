import * as React from 'react';
import { MainState } from './Main';

export interface FormProps extends MainState {
    onToggleState(stateKey: string): void,
    onInputChange(stateKey: string, event: React.SyntheticEvent<HTMLInputElement>): void,
    onSubmit(event: React.SyntheticEvent<HTMLFormElement>): void,
}
