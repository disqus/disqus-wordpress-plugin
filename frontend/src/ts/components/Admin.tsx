import * as React from 'react';
import * as ReactRedux from 'react-redux';
import * as Redux from 'redux';
import {
    ExportCommentsContainer,
    InstallContainer,
    SiteConfigContainer,
    SSOConfigContainer,
    SupportDiagnosticsContainer,
    SyncConfigContainer,
} from '../containers';
import { IAdminState } from '../reducers/AdminState';
import { getForumAdminUrl, getWordpressAdminUrl } from '../utils';
import { IFormProps } from './FormProps';
import SupportLinks from './SupportLinks';
import WelcomePanel from './WelcomePanel';

const getSSOContainer = (props: IFormProps) => {
    const adminOptions = props.data.adminOptions;
    if (!adminOptions.disqus_public_key || !adminOptions.disqus_secret_key || !adminOptions.disqus_installed) {
        return (
            <div className='notice notice-warning'>
                <p>
                    <span className='dashicons dashicons-warning' />
                    {' '}
                    {/* tslint:disable-next-line:max-line-length */}
                    {__('You must have a Site Shortname, API Public Key, and API Secret Key configured to enable this feature.')}
                </p>
            </div>
        );
    }
    return <SSOConfigContainer />;
};

const getSyncContainer = (props: IFormProps) => {
    const adminOptions = props.data.adminOptions;
    if (!adminOptions.disqus_secret_key || !adminOptions.disqus_admin_access_token || !adminOptions.disqus_installed) {
        return (
            <div className='notice notice-warning'>
                <p>
                    <span className='dashicons dashicons-warning' />
                    {' '}
                    {/* tslint:disable-next-line:max-line-length */}
                    {__('You must have a Site Shortname, API Secret Key, and API Access Token configured to enable this feature.')}
                </p>
            </div>
        );
    }

    return <SyncConfigContainer />;
};

const getActiveTab = (props: IFormProps) => (
    props.data.activeTab || (props.data.adminOptions.disqus_installed ? 'siteConfiguration' : 'install')
);

const getExportContainer = (props: IFormProps) => {
    const adminOptions = props.data.adminOptions;
    if (!adminOptions.disqus_secret_key || !adminOptions.disqus_admin_access_token) {
        return (
            <div className='notice notice-warning'>
                <p>
                    <span className='dashicons dashicons-warning' />
                    {' '}
                    {__('You must have an API Secret Key and API Access Token configured to enable this feature.')}
                </p>
            </div>
        );
    }

    return <ExportCommentsContainer />;
};

const getTabClassName = (props: IFormProps, id: string) => {
    const activeTab = getActiveTab(props);
    return `nav-tab${activeTab === id ? ' nav-tab-active' : ''}`;
};

const AdminTabBar = (props: IFormProps) => (
    <div className='nav-tab-wrapper'>
        <a href='#install' className={getTabClassName(props, 'install')}>
            {props.data.adminOptions.disqus_installed ? __('Reinstall') : __('Install')}
        </a>
        <a href='#siteConfiguration' className={getTabClassName(props, 'siteConfiguration')}>
            {__('Site Configuration')}
        </a>
        <a href='#syncing' className={getTabClassName(props, 'syncing')}>
            {__('Syncing')}
        </a>
        <a href='#singleSignOn' className={getTabClassName(props, 'singleSignOn')}>
            {__('Single Sign-on')}
        </a>
        <a href='#support' className={getTabClassName(props, 'support')}>
            {__('Support')}
        </a>
    </div>
);

/* tslint:disable:max-line-length */
const getActiveTabView = (props: IFormProps) => {
    const activeTab = getActiveTab(props);
    switch (activeTab) {
    case 'syncing':
        return (
            <div>
                <h3>
                    {__('WordPress Comments')}
                </h3>
                <p className='description'>
                    {__('Disqus has replaced the default WordPress commenting system. You may access and edit the comments in your database, but any actions performed there will not be reflected in Disqus.')}
                </p>
                <p className='submit'>
                    <a
                        href={getWordpressAdminUrl('editComments')}
                        className='button button-large'
                    >
                        <span className='dashicons dashicons-wordpress-alt' />
                        {' '}
                        {__('View WordPress Comments')}
                    </a>
                </p>
                <hr />
                <h3>
                    {__('Sync')}
                </h3>
                <p className='description'>
                    {__('Copy comments created and edited in Disqus to your local WordPress database for backup purposes.')}
                </p>
                {getSyncContainer(props)}
                <hr />
                <h3>
                    {__('Export')}
                </h3>
                <p className='description'>
                    {__('Export comments from your WordPress installation to Disqus.')}
                </p>
                {getExportContainer(props)}
            </div>
        );
    case 'singleSignOn':
        return (
            <div>
                <h3>
                    {__('Set up Single Sign-on')}
                </h3>
                <p className='description'>
                    {__('Allow users to sign in with this site\'s user accounts. This is a Disqus Pro feature.')}
                    {' '}
                    <a href='https://help.disqus.com/customer/portal/articles/1148635' target='_blank'>
                        {__('Learn More')}
                    </a>
                </p>
                {getSSOContainer(props)}
            </div>
        );
    case 'support':
        return (
            <div>
                <h3>
                    {__('Support Links')}
                </h3>
                <SupportLinks />
                <hr />
                <h3>
                    {__('Diagnostic Information')}
                </h3>
                <p className='description'>
                    {__('Include the following information in any private support requests, but do not share this publicly.')}
                </p>
                <div className='submit'>
                    <SupportDiagnosticsContainer />
                </div>
            </div>
        );
    case 'install':
        return <InstallContainer />;
    case 'siteConfiguration':
    default:
        return (
            <div>
                <h3>
                    {__('Site Configuration')}
                </h3>
                <p className='description'>
                    {__('Your site configuration comes from Disqus. Changing these values may break your installation.')}
                    {' '}
                    <a
                        href={getForumAdminUrl(props.data.adminOptions.disqus_forum_url, 'install/platforms/wordpress')}
                        target='_blank'
                    >
                        {__('WordPress install instructions')}
                    </a>
                </p>
                <SiteConfigContainer />
            </div>
        );
    }
};

const Admin = (props: IFormProps) => (
    <div>
        {props.data.adminOptions.disqus_installed ? <WelcomePanel shortname={props.data.adminOptions.disqus_forum_url} /> : null}
        <AdminTabBar {...props} />
        {getActiveTabView(props)}
    </div>
);
/* tslint:enable:max-line-length */

export default Admin;
