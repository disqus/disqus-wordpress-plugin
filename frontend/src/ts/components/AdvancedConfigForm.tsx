import * as React from 'react';
import { IFormProps } from './FormProps';

/* tslint:disable:max-line-length */
const AdvancedConfigForm = (props: IFormProps) => (
    <form name='advanced' action='' method='POST' onSubmit={props.onSubmitSiteForm}>
        <table className='form-table'>
            <tbody>
                <tr>
                    <th scope='row'>
                        <label htmlFor='disqus_render_js'>
                            {__('Render Comments JavaScript')}
                        </label>
                    </th>
                    <td>
                        <input
                            type='checkbox'
                            id='disqus_render_js'
                            name='disqus_render_js'
                            checked={Boolean(props.data.localAdminOptions.get('disqus_render_js'))}
                            onChange={props.onInputChange.bind(null, 'disqus_render_js')}
                        />
                        <p className='description'>
                            {__('This will render the Disqus comments javascript directly into the page markup, rather than use the wp_enqueue_script() function. Enable this if Disqus doesn\'t load on your posts.')}
                        </p>
                    </td>
                </tr>
            </tbody>
        </table>
        <p className='submit'>
            <input
                type='submit'
                name='submit-application-form'
                className='button button-primary'
                value={__('Save')}
            />
        </p>
    </form>
);
/* tslint:enable:max-line-length */

export default AdvancedConfigForm;
