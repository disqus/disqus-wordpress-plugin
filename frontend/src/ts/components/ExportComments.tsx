import * as React from 'react';
import { ExportLogStaus, IExportPostLog } from '../reducers/AdminState';
import { IFormProps } from './FormProps';

const getStatusMessage = (props: IFormProps): JSX.Element => {
    if (props.data.isExportRunning)
        return <h4>{__('Exporting') + 'â€¦'}</h4>;

    return (
        <h4>
            <span>
                {__('Done!')}
                {' '}
                <a href={`https://import.disqus.com/${props.data.adminOptions.disqus_forum_url}/`} target='_blank'>
                    {__('Check your import status')}
                </a>
            </span>
        </h4>
    );
};

const LogMessages = (props: IFormProps) => {
    const logArray: IExportPostLog[] = props.data.exportLogs.toArray();
    const logElements: JSX.Element[] = logArray.map((log: IExportPostLog) => {
        let statusColor: string;
        let statusText: JSX.Element|string;
        switch (log.status) {
        case ExportLogStaus.failed:
            statusColor = 'red';
            statusText = <abbr title={log.error}>{__('Failed')}</abbr>;
            break;
        case ExportLogStaus.complete:
            statusColor = 'green';
            statusText = __('Complete');
            break;
        case ExportLogStaus.pending:
        default:
            statusColor = 'gray';
            statusText = __('Pending');
            break;
        }

        return (
            <tr key={log.id}>
                <td>{log.id}</td>
                <td><a href={log.link}>{log.title}</a></td>
                <td><strong style={{ color: statusColor }}>{statusText}</strong></td>
                <td>{log.numComments}</td>
            </tr>
        );
    });

    return (
        <div>
            {getStatusMessage(props)}
            <table className='wp-list-table widefat striped'>
                <thead>
                    <tr>
                        <th scope='col'>{__('ID')}</th>
                        <th scope='col'>{__('Title')}</th>
                        <th scope='col'>{__('Status')}</th>
                        <th scope='col'>{__('# Imported')}</th>
                    </tr>
                </thead>
                <tbody>
                    {logElements}
                </tbody>
            </table>
        </div>
    );
};

const ExportComments = (props: IFormProps) => (
    <form
        name='export'
        method='POST'
        onSubmit={props.onSubmitExportCommentsForm}
    >
        <p className='submit'>
            <button type='submit' className='button button-large' disabled={props.data.isExportRunning}>
                <span className='dashicons dashicons-upload' />
                {' '}
                {__('Export Comments')}
            </button>
        </p>
        {props.data.exportLogs.size ? <LogMessages {...props} /> : null}
    </form>
);

export default ExportComments;
