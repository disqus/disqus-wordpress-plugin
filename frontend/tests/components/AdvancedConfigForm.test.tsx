import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import AdvancedConfigForm from '../../src/ts/components/AdvancedConfigForm';
import { IFormProps } from '../../src/ts/components/FormProps';
import { getDefaultFormProps } from '../fixtures';

describe('AdvancedConfigForm component', () => {
    test('Renders basic form', () => {
        const props: IFormProps = getDefaultFormProps();
        const component = TestRenderer.create(<AdvancedConfigForm {...props} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Renders form with values', () => {
        const props: IFormProps = getDefaultFormProps();
        props.data = props.data.with({
            localAdminOptions: props.data.localAdminOptions.with({
                disqus_render_js: true,
            }),
        });
        const component = TestRenderer.create(<AdvancedConfigForm {...props} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    test('onInputChange prop gets called when values change', () => {
        const props: IFormProps = getDefaultFormProps();
        const onInputChange = props.onInputChange = jest.fn();
        props.data = props.data.with({
            localAdminOptions: props.data.localAdminOptions.with({
                disqus_render_js: true,
            }),
        });
        const component = TestRenderer.create(<AdvancedConfigForm {...props} />);
        const root = component.root;

        root.findAllByType('input').forEach((input: any) => {
            if (input.props.type !== 'submit') {
                input.props.onChange.call(null, { foo: 'bar' });
                expect(onInputChange).toHaveBeenLastCalledWith(input.props.id, { foo: 'bar' });
            }
        });
    });

    test('Submitting the form calls onSubmitSiteForm prop', () => {
        const props: IFormProps = getDefaultFormProps();
        const onSubmitSiteForm = props.onSubmitSiteForm = jest.fn();
        props.data = props.data.with({
            localAdminOptions: props.data.localAdminOptions.with({
                disqus_render_js: true,
            }),
        });
        const component = TestRenderer.create(<AdvancedConfigForm {...props} />);
        const root = component.root;

        const form = root.findByType('form');
        form.props.onSubmit();

        expect(onSubmitSiteForm).toHaveBeenCalled();
    });
});
