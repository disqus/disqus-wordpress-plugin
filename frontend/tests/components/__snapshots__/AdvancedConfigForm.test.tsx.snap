// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedConfigForm component Renders basic form 1`] = `
<form
  action=""
  method="POST"
  name="advanced"
  onSubmit={[MockFunction]}
>
  <table
    className="form-table"
  >
    <tbody>
      <tr>
        <th
          scope="row"
        >
          <label
            htmlFor="disqus_render_js"
          >
            Render Comments JavaScript
          </label>
        </th>
        <td>
          <input
            checked={false}
            id="disqus_render_js"
            name="disqus_render_js"
            onChange={[Function]}
            type="checkbox"
          />
          <p
            className="description"
          >
            This will render the Disqus comments javascript directly into the page markup, rather than use the wp_enqueue_script() function. Enable this if Disqus doesn't load on your posts.
          </p>
        </td>
      </tr>
    </tbody>
  </table>
  <p
    className="submit"
  >
    <input
      className="button button-primary"
      name="submit-application-form"
      type="submit"
      value="Save"
    />
  </p>
</form>
`;

exports[`AdvancedConfigForm component Renders form with values 1`] = `
<form
  action=""
  method="POST"
  name="advanced"
  onSubmit={[MockFunction]}
>
  <table
    className="form-table"
  >
    <tbody>
      <tr>
        <th
          scope="row"
        >
          <label
            htmlFor="disqus_render_js"
          >
            Render Comments JavaScript
          </label>
        </th>
        <td>
          <input
            checked={true}
            id="disqus_render_js"
            name="disqus_render_js"
            onChange={[Function]}
            type="checkbox"
          />
          <p
            className="description"
          >
            This will render the Disqus comments javascript directly into the page markup, rather than use the wp_enqueue_script() function. Enable this if Disqus doesn't load on your posts.
          </p>
        </td>
      </tr>
    </tbody>
  </table>
  <p
    className="submit"
  >
    <input
      className="button button-primary"
      name="submit-application-form"
      type="submit"
      value="Save"
    />
  </p>
</form>
`;
